bitpanda - websockets

pcriadoperez opened the pull request:
- Added bitpanda WS implementation
- Was recommended to also merge these changes https://github.com/kroitor/ccxt.pro/pull/346 because we also have the need to re-subscribe the previous markets in some streams

fixes: #17572

pcriadoperez commented:
### watchTicker
```python
bitpanda.watchTicker(BTC/USDT)
{'ask': None,
 'askVolume': None,
 'average': None,
 'baseVolume': None,
 'bid': None,
 'bidVolume': None,
 'change': 218.48,
 'close': 21946.27,
 'datetime': '2023-02-12T14:45:43.756155Z',
 'high': 21987.05,
 'info': {'high': '21987.05',
          'instrument': 'BTC_USDT',
          'last_price': '21946.27',
          'low': '21628.23',
          'price_change': '218.48',
          'price_change_percentage': '1.01',
          'price_points': [{'price': '21727.79',
                            'time': '2023-02-11T14:43:32.299785Z'},
                           {'price': '21687.43',
                            'time': '2023-02-11T16:59:59.999999Z'},
                           {'price': '21628.23',
                            'time': '2023-02-11T18:59:59.999999Z'},
                           {'price': '21659.3',
                            'time': '2023-02-11T19:59:59.999999Z'},
                           {'price': '21703.16',
                            'time': '2023-02-11T20:59:59.999999Z'},
                           {'price': '21817.7',
                            'time': '2023-02-11T21:59:59.999999Z'},
                           {'price': '21837.8',
                            'time': '2023-02-11T22:59:59.999999Z'},
                           {'price': '21866.7',
                            'time': '2023-02-11T23:59:59.999999Z'},
                           {'price': '21777.98',
                            'time': '2023-02-12T04:59:59.999999Z'},
                           {'price': '21805.78',
                            'time': '2023-02-12T06:59:59.999999Z'},
                           {'price': '21799.61',
                            'time': '2023-02-12T07:59:59.999999Z'},
                           {'price': '21849.02',
                            'time': '2023-02-12T08:59:59.999999Z'},
                           {'price': '21915.59',
                            'time': '2023-02-12T09:59:59.999999Z'},
                           {'price': '21876.68',
                            'time': '2023-02-12T10:59:59.999999Z'},
                           {'price': '21860.54',
                            'time': '2023-02-12T12:59:59.999999Z'},
                           {'price': '21800.02',
                            'time': '2023-02-12T13:59:59.999999Z'},
                           {'price': '21946.27',
                            'time': '2023-02-12T14:44:56.614460Z'}],
          'volume': '61854.587917'},
 'last': 21946.27,
 'low': 21628.23,
 'open': 21727.79,
 'percentage': 1.01,
 'previousClose': None,
 'quoteVolume': 61854.587917,
 'symbol': 'BTC/USDT',
 'timestamp': 1676213143756,
 'vwap': None}
 ```
 ### watchTickers
 ```python
 bitpanda.watchTickers()
{'1INCH/EUR': {'ask': None,
               'askVolume': None,
               'average': None,
               'baseVolume': None,
               'bid': None,
               'bidVolume': None,
               'change': 0.0,
               'close': 0.497,
               'datetime': '2023-02-12T14:51:15.658458Z',
               'high': 0.497,
               'info': {'high': '0.497',
                        'instrument': '1INCH_EUR',
                        'last_price': '0.497',
                        'low': '0.497',
                        'price_change': '0.0',
                        'price_change_percentage': '0.0',
                        'price_points': [],
                        'volume': '0.0'},
               'last': 0.497,
               'low': 0.497,
               'open': 0.497,
               'percentage': 0.0,
               'previousClose': None,
               'quoteVolume': 0.0,
               'symbol': '1INCH/EUR',
               'timestamp': 1676213475658,
               'vwap': None},
 'AAVE/EUR': {'ask': None,
              'askVolume': None,
              'average': None,
              'baseVolume': None,
              'bid': None,
              'bidVolume': None,
              'change': 0.97,
              'close': 74.27,
              'datetime': '2023-02-12T14:51:15.658458Z',
              'high': 74.9,
              'info': {'high': '74.9',
                       'instrument': 'AAVE_EUR',
                       'last_price': '74.27',
                       'low': '73.28',
                       'price_change': '0.97',
                       'price_change_percentage': '1.32',
                       'price_points': [{'price': '73.3',
                                         'time': '2023-02-11T14:50:59.009535Z'},
                                        {'price': '73.28',
                                         'time': '2023-02-11T17:59:59.999999Z'},
                                        {'price': '74.17',
                                         'time': '2023-02-11T22:59:59.999999Z'},
                                        {'price': '73.99',
                                         'time': '2023-02-12T00:59:59.999999Z'},
                                        {'price': '73.99',
                                         'time': '2023-02-12T02:59:59.999999Z'},
                                        {'price': '74.21',
                                         'time': '2023-02-12T06:59:59.999999Z'},
                                        {'price': '74.62',
                                         'time': '2023-02-12T08:59:59.999999Z'},
                                        {'price': '74.51',
                                         'time': '2023-02-12T10:59:59.999999Z'},
                                        {'price': '74.6',
                                         'time': '2023-02-12T12:59:59.999999Z'},
                                        {'price': '74.27',
                                         'time': '2023-02-12T13:59:59.999999Z'}],
                       'volume': '19579.3952004'},
              'last': 74.27,
              'low': 73.28,
              'open': 73.3,
              'percentage': 1.32,
              'previousClose': None,
              'quoteVolume': 19579.3952004,
              'symbol': 'AAVE/EUR',
              'timestamp': 1676213475658,
              'vwap': None},
              ...
```
### watchOrderBook
```python
bitpanda.watchOrderBook(BTC/USDT)
{'asks': [[21908.55, 0.04108], [21908.58, 0.04108], [21908.63, 0.02281], [21912.42, 0.09126], [21914.6, 0.22812], [21916.19, 0.01984], [21919.51, 0.22807], [21921.34, 0.22805], [21928.9, 0.0912], [21931.75, 0.45589], [21932.37, 0.19997], [22027.11, 0.45392], [22142.48, 0.24833], [22377.32, 0.49958], [22604.82, 1.48539], [22832.42, 1.98287], [23370.0, 0.00177], [23400.0, 0.00441], [23987.56, 0.04627], [26832.45, 0.00908], [28832.45, 0.00709], [29832.45, 0.00547], [36574.0, 0.002], [36576.0, 0.05506], [38096.66, 0.023], [80155.0, 0.00044], [80156.0, 0.06915], [80169.85, 0.05], [890000.0, 0.01439], [999999.0, 0.0049], [9999998.99, 0.01007], [9999999.0, 0.0071]],
 'bids': [[21905.55, 0.02282], [21905.32, 0.04108], [21905.31, 0.04108], [21901.86, 0.0913], [21899.18, 0.22828], [21898.67, 0.01123], [21896.97, 0.22831], [21890.6, 0.22837], [21885.09, 0.0791], [21881.76, 0.19892], [21880.69, 0.45696], [21788.14, 0.4589], [21750.02, 0.00459], [21700.02, 0.0046], [21679.23, 0.2496], [21600.02, 0.00462], [21500.0, 0.00465], [21458.26, 0.49554], [21300.41, 0.00463], [21219.99, 1.48757], [21140.0, 0.04731], [21003.02, 1.99601], [21000.0, 0.00476], [20850.0, 0.0096], [20520.68, 0.0034], [20500.0, 0.09895], [20000.0, 0.025], [19721.0, 0.00067], [19720.68, 0.00359], [19500.0, 0.05129], [19000.0, 0.05264], [18500.0, 0.06049], [17000.0, 0.02572], [10100.0, 0.24683], [10071.0, 0.02977], [10000.0, 0.02], [3605.0, 0.64331], [2881.19, 0.03694], [1213.18, 0.16469], [1169.95, 0.3575], [1169.85, 0.19761], [556.0, 0.22605], [158.02, 0.981], [158.01, 1.10654], [156.0, 2.03029], [155.19, 0.34981], [4.28, 9.34813], [4.27, 22.02318], [4.23, 11.6911], [3.29, 6.58228], [2.2, 4.54546], [0.01, 1000.0]],
 'datetime': '2023-02-12T14:52:19.001361Z',
 'nonce': 1676213539001,
 'symbol': 'BTC/USDT',
 'timestamp': 1676213539001}
 ```
 ### watchOHLCV
 ```python
 bitpanda.watchOHLCV(BTC/USDT)
[[1676215079999, 21957.99, 21957.99, 21957.99, 21957.99, 499.7638524]]
```
 ### watchBalance
 ```python
bitpanda.watchBalance()
{'BTC': {'free': 0.0, 'total': 0.0, 'used': 0.0},
 'EUR': {'free': 0.0080343, 'total': 0.0080343, 'used': 0.0},
 'free': {'BTC': 0.0, 'EUR': 0.0080343},
 'info': {'account_id': '49202c1a-48dc-423e-b336-bb65baccc7bd',
          'balances': [{'account_holder': '49202c1a-48dc-423e-b336-bb65baccc7bd',
                        'account_id': '49202c1a-48dc-423e-b336-bb65baccc7bd',
                        'available': '0.0',
                        'change': '0.00024853',
                        'currency_code': 'BTC',
                        'locked': '0.0',
                        'sequence': 9688440405,
                        'time': '2022-10-12T00:13:26.235131Z'},
                       {'account_holder': '49202c1a-48dc-423e-b336-bb65baccc7bd',
                        'account_id': '49202c1a-48dc-423e-b336-bb65baccc7bd',
                        'available': '0.0080343',
                        'change': '12.98',
                        'currency_code': 'EUR',
                        'locked': '0.0',
                        'sequence': 9688440399,
                        'time': '2022-10-12T00:13:26.185386Z'}],
          'channel_name': 'ACCOUNT_HISTORY',
          'time': '2023-02-12T14:58:42.866499Z',
          'type': 'BALANCES_SNAPSHOT'},
 'total': {'BTC': 0.0, 'EUR': 0.0080343},
 'used': {'BTC': 0.0, 'EUR': 0.0}}
 ```
 ### watchOrders
 ```python
 bitpanda.watchOrders()
[{'amount': 0.00053,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': '2023-02-12T15:06:05.034Z',
  'fees': [],
  'filled': 0.0,
  'id': 'a7721430-3227-46a1-80fa-fe74b78139c3',
  'info': {'order': {'account_holder': '49202c1a-48dc-423e-b336-bb65baccc7bd',
                     'account_id': '49202c1a-48dc-423e-b336-bb65baccc7bd',
                     'amount': '0.00053',
                     'filled_amount': '0.0',
                     'instrument_code': 'BTC_EUR',
                     'is_post_only': False,
                     'order_id': 'a7721430-3227-46a1-80fa-fe74b78139c3',
                     'price': '22548.99',
                     'sequence': 10807434347,
                     'side': 'SELL',
                     'status': 'OPEN',
                     'time': '2023-02-12T15:06:05.034023Z',
                     'time_in_force': 'GOOD_TILL_CANCELLED',
                     'type': 'LIMIT'},
           'trades': []},
  'lastTradeTimestamp': None,
  'postOnly': False,
  'price': 22548.99,
  'remaining': 0.00053,
  'side': 'sell',
  'status': 'open',
  'stopPrice': None,
  'symbol': 'BTC/EUR',
  'timeInForce': 'GTC',
  'timestamp': 1676214365034,
  'trades': [],
  'type': 'limit'}]
  ```
  ### watchMyTrades
  ```python
  bitpanda.watchMyTrades()
[{'amount': 0.00086,
  'cost': 17.5204016,
  'datetime': '2023-02-12T21:56:51.042Z',
  'fee': {'cost': 0.0, 'currency': 'BTC', 'rate': 0.0},
  'fees': [{'cost': 0.0, 'currency': 'BTC', 'rate': 0.0}],
  'id': 'a44b312f-d449-44cd-9044-1414f6e9e160',
  'info': {'account_holder': '49202c1a-48dc-423e-b336-bb65baccc7bd',
           'account_id': '49202c1a-48dc-423e-b336-bb65baccc7bd',
           'amount': '0.00086',
           'instrument_code': 'BTC_EUR',
           'order_id': '853a1479-7c61-4a5e-8ff5-f76fcd0806ab',
           'price': '20372.56',
           'price_tick_sequence': 0,
           'side': 'BUY',
           'time': '2023-02-12T21:56:51.042889Z',
           'trade_id': 'a44b312f-d449-44cd-9044-1414f6e9e160'},
  'order': '853a1479-7c61-4a5e-8ff5-f76fcd0806ab',
  'price': 20372.56,
  'side': 'buy',
  'symbol': 'BTC/EUR',
  'takerOrMaker': 'taker',
  'timestamp': 1676239011042,
  'type': None}]
```

carlosmiei commented:
> Was recommended to also merge these changes https://github.com/kroitor/ccxt.pro/pull/346 because we also have the need to re-subscribe the previous markets in some streams
@pcriadoperez  is this still necessary?

pcriadoperez commented:
> > Was recommended to also merge these changes [kroitor/ccxt.pro#346](https://github.com/kroitor/ccxt.pro/pull/346) because we also have the need to re-subscribe the previous markets in some streams
> 
> @pcriadoperez is this still necessary?

I had to test it again because it had been a while. But yes, without the other PR if you connect to several symbols simultanesult it will fail.

What happens is:
- User subscribe to one symbol
- Bitpanda calls SUBSCRIBE for the first symbol
- User subscribe to the second symbol
- It checks the subscription to the orderbook channel and sends an UPDATE_SUBSCRIPTION with the first symbol and the second symbol
- If these calls happen very close to each other it happens that the subscription hasn´t been created yet as with the current implementation it waits for the client to be connected, and fails to read the subscription and update it correctly.

I closed the pr that was on the old ccxt.pro repo and moved it to this one and added also some changes for error handling.
#17723 

As an alternative, if we don´t want to edit the global exchange files, I could try to handle it locally by storing the subscriptions in the options of bitpanda.. however I felt that would be more of a patch just for bitpanda so it was better solve globally

carlosmiei commented on ts/src/pro/bitpanda.ts:
```diff
@@ -0,0 +1,1335 @@
+        if (client !== undefined) {
+            subscription = this.safeValue (client.subscriptions, subscriptionHash);
+            if (subscription !== undefined) {
+                const ohlcvMarket = this.safeValue (subscription, marketId, {});
+                const marketSubscribed = this.safeValue (ohlcvMarket, timeframe, false);
+                if (!marketSubscribed) {
+                    type = 'UPDATE_SUBSCRIPTION';
+                    client.subscriptions[subscriptionHash] = undefined;
+                }
+            } else {
+                subscription = {};
+            }
+        }
+        const subscriptionMarketId = this.safeValue (subscription, marketId);
+        if (subscriptionMarketId === undefined) {
+            subscription[marketId] = {};
+        }
+        subscription[marketId][timeframe] = true;
+        const properties = [];
+        const marketIds = Object.keys (subscription);
+        for (let i = 0; i < marketIds.length; i++) {
+            const marketIdtimeframes = Object.keys (subscription[marketIds[i]]);
+            for (let ii = 0; ii < marketIdtimeframes.length; ii++) {
+                const marketTimeframeId = this.safeValue (timeframes, timeframe);
+                const property = {
+                    'instrument_code': marketIds[i],
+                    'time_granularity': marketTimeframeId,
+                };
+                properties.push (property);
+            }
+        }
+        const request = {
+            'type': type,
+            'channels': [
+                {
+                    'name': 'CANDLESTICKS',
+                    'properties': properties,
+                },
+            ],
+        };
```
question, can't this logic be abstracted using `watchMultiple` like you did with other methods?

pcriadoperez replied:
I looked into it.. but as we look both at the timeframe and marketId it´s a bit different than the other watchMultiple. And seemed more complicated to have one function do everything than keep it seperate even if the code is repeated... decided to keep it KISS over DRY 🤣

carlosmiei commented on ts/src/pro/bitpanda.ts:
```diff
@@ -0,0 +1,1335 @@
+         * @param {object} params extra parameters specific to the bitpanda api endpoint
+         * @returns {[[int]]} A list of candles ordered as timestamp, open, high, low, close, volume
+         */
+        await this.loadMarkets ();
```
@pcriadoperez for some reason bitpand is not broadcasting candles
```
2023-05-15T11:46:14.863Z connecting to wss://streams.exchange.bitpanda.com with timeout 10000 ms
2023-05-15T11:46:15.079Z connected
2023-05-15T11:46:15.079Z ping loop
2023-05-15T11:46:15.080Z sending {'type': 'SUBSCRIBE', 'channels': [{'name': 'CANDLESTICKS', 'properties': [{'instrument_code': 'BTC_USDT', 'time_granularity': {'unit': 'MINUTES', 'period': 1}}]}]}
2023-05-15T11:46:15.132Z pong WSMessage(type=<WSMsgType.PONG: 10>, data=bytearray(b''), extra='')
2023-05-15T11:46:15.132Z message {"channels":[{"properties":[{"instrument_code":"BTC_USDT","time_granularity":{"unit":"MINUTES","period":1}}],"name":"CANDLESTICKS"}],"type":"SUBSCRIPTIONS","time":"2023-05-15T11:46:15.143408Z"}
2023-05-15T11:46:15.380Z sending {'type': 'UPDATE_SUBSCRIPTION', 'channels': [{'name': 'CANDLESTICKS', 'properties': [{'instrument_code': 'BTC_USDT', 'time_granularity': {'unit': 'MINUTES', 'period': 1}}, {'instrument_code': 'LTC_EUR', 'time_granularity': {'unit': 'MINUTES', 'period': 1}}]}]}
2023-05-15T11:46:15.436Z message {"channels":[{"properties":[{"instrument_code":"BTC_USDT","time_granularity":{"unit":"MINUTES","period":1}},{"instrument_code":"LTC_EUR","time_granularity":{"unit":"MINUTES","period":1}}],"name":"CANDLESTICKS"}],"type":"SUBSCRIPTIONS","time":"2023-05-15T11:46:15.445270Z"}
2023-05-15T11:46:15.436Z message {"channel_name":"CANDLESTICKS","type":"SUBSCRIPTION_UPDATED","time":"2023-05-15T11:46:15.445366Z"}
2023-05-15T11:46:25.080Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:46:25.090120Z"}
2023-05-15T11:46:35.149Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:46:35.089555Z"}
2023-05-15T11:46:45.079Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:46:45.090015Z"}
2023-05-15T11:46:45.132Z pong WSMessage(type=<WSMsgType.PONG: 10>, data=bytearray(b''), extra='')
2023-05-15T11:46:55.120Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:46:55.089517Z"}
2023-05-15T11:47:05.086Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:47:05.089662Z"}
2023-05-15T11:47:15.092Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:47:15.090249Z"}
2023-05-15T11:47:15.141Z pong WSMessage(type=<WSMsgType.PONG: 10>, data=bytearray(b''), extra='')
2023-05-15T11:47:25.081Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:47:25.089497Z"}
2023-05-15T11:47:35.079Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:47:35.089340Z"}
2023-05-15T11:47:45.079Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:47:45.089621Z"}
2023-05-15T11:47:45.144Z pong WSMessage(type=<WSMsgType.PONG: 10>, data=bytearray(b''), extra='')
2023-05-15T11:47:55.123Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:47:55.090032Z"}
2023-05-15T11:48:05.161Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:48:05.089720Z"}
2023-05-15T11:48:15.099Z message {"subscription":"SYSTEM","channel_name":"SYSTEM","type":"HEARTBEAT","time":"2023-05-15T11:48:15.089322Z"}
2023-05-15T11:48:15.155Z pong WSMessage(type=<WSMsgType.PONG: 10>, data=bytearray(b''), extra='')
```

pcriadoperez replied:
it's very strange, it works for me about 2/10 times

carlosmiei replied:
@pcriadoperez now I got a different error 
```
2023-05-16T09:01:49.923Z connecting to wss://streams.exchange.bitpanda.com with timeout 10000 ms
2023-05-16T09:01:50.139Z connected
2023-05-16T09:01:50.139Z ping loop
2023-05-16T09:01:50.139Z sending {'type': 'SUBSCRIBE', 'channels': [{'name': 'CANDLESTICKS', 'properties': [{'instrument_code': 'BTC_USDT', 'time_granularity': {'unit': 'MINUTES', 'period': 1}}]}]}
2023-05-16T09:01:50.193Z pong WSMessage(type=<WSMsgType.PONG: 10>, data=bytearray(b''), extra='')
2023-05-16T09:01:50.193Z message {"channels":[{"properties":[{"instrument_code":"BTC_USDT","time_granularity":{"unit":"MINUTES","period":1}}],"name":"CANDLESTICKS"}],"type":"SUBSCRIPTIONS","time":"2023-05-16T09:01:50.213721Z"}
2023-05-16T09:01:50.439Z sending {'type': 'UPDATE_SUBSCRIPTION', 'channels': [{'name': 'CANDLESTICKS', 'properties': [{'instrument_code': 'BTC_USDT', 'time_granularity': {'unit': 'MINUTES', 'period': 1}}, {'instrument_code': 'BTC_EUR', 'time_granularity': {'unit': 'MINUTES', 'period': 1}}]}]}
2023-05-16T09:01:50.492Z message {"channels":[{"properties":[{"instrument_code":"BTC_USDT","time_granularity":{"unit":"MINUTES","period":1}},{"instrument_code":"BTC_EUR","time_granularity":{"unit":"MINUTES","period":1}}],"name":"CANDLESTICKS"}],"type":"SUBSCRIPTIONS","time":"2023-05-16T09:01:50.510524Z"}
2023-05-16T09:01:50.492Z message {"channel_name":"CANDLESTICKS","type":"SUBSCRIPTION_UPDATED","time":"2023-05-16T09:01:50.510592Z"}
2023-05-16T09:01:50.510Z message {"instrument_code":"BTC_EUR","granularity":{"unit":"MINUTES","period":1},"high":"24974.5","low":"24974.5","open":"24974.5","close":"24974.5","volume":"12.48725","last_sequence":2864885,"channel_name":"CANDLESTICKS","type":"CANDLESTICK_SNAPSHOT","time":"2023-05-16T09:01:59.999999Z"}
Exchange.filter_by_since_limit() takes from 2 to 5 positional arguments but 6 were given
```

carlosmiei replied:
nvm I had an outdated version apparently

carlosmiei commented on ts/src/pro/bitpanda.ts:
```diff
@@ -0,0 +1,1335 @@
+        }
+        await this.authenticate (params);
+        const url = this.urls['api']['ws'];
+        const messageHash = 'myTrades';
```
@pcriadoperez why aren't we using the symbol to build the messageHash? 
how could we resolve two different watchMyTrades futures without it?
```
a = bitpanda.watchMyTrades('LTC/EUR')
b =  bitpanda.watchMyTrades('BTC/USDT')
```

pcriadoperez replied:
Great catch! I added a fix

carlosmiei commented on ts/src/pro/bitpanda.ts:
```diff
@@ -0,0 +1,1335 @@
+            const market = this.market (symbol);
+            symbol = market['symbol'];
+        }
+        await this.authenticate (params);
```
same question here

carlosmiei commented on ts/src/pro/bitpanda.ts:
```diff
@@ -0,0 +1,1360 @@
+        const url = this.urls['api']['ws'];
+        const messageHash = 'balance';
+        const subscribeHash = 'ACCOUNT_HISTORY';
+        const bpRemainingQuota = this.safeNumber (this.options, 'bp_remaining_quota', 200);
```
`safeNumber` might potentially return a string, so we should use `safeInteger` instead (remaining requests seems to be an integer)

carlosmiei commented on ts/src/pro/bitpanda.ts:
```diff
@@ -0,0 +1,1361 @@
+        const bidAsk = this.parseBidAsk (delta, 1, 2);
+        const type = this.safeString (delta, 0);
+        if (type === 'BUY') {
+            orderbook['bids'].storeArray (bidAsk);
```
also this is not working on php, taking a look

carlosmiei commented:
LGTM ✅ 


```
p bitpanda watchBalance          
Python v3.10.9
CCXT v3.0.104
bitpanda.watchBalance()
{'LTC': {'free': 0.0, 'total': 0.0, 'used': 0.0},
 'USDT': {'free': 34.144, 'total': 34.144, 'used': 0.0},
 'free': {'LTC': 0.0, 'USDT': 34.144},
 'info': {'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
          'balances': [{'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'available': '34.144',
                        'change': '34.144',
                        'currency_code': 'USDT',
                        'locked': '0.0',
                        'sequence': 11799592776,
                        'time': '2023-05-15T11:43:03.561608Z'},
                       {'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'available': '0.0',
                        'change': '1.16178',
                        'currency_code': 'LTC',
                        'locked': '0.0',
                        'sequence': 9688664235,
                        'time': '2022-10-12T00:56:19.697459Z'}],
          'channel_name': 'ACCOUNT_HISTORY',
          'time': '2023-05-16T09:45:22.571643Z',
          'type': 'BALANCES_SNAPSHOT'},
 'total': {'LTC': 0.0, 'USDT': 34.144},
 'used': {'LTC': 0.0, 'USDT': 0.0}}
{'LTC': {'free': 0.0, 'total': 0.0, 'used': 0.0},
 'USDT': {'free': 24.053528, 'total': 34.144, 'used': 10.090472},
 'free': {'LTC': 0.0, 'USDT': 24.053528},
 'info': {'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
          'balances': [{'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'available': '34.144',
                        'change': '34.144',
                        'currency_code': 'USDT',
                        'locked': '0.0',
                        'sequence': 11799592776,
                        'time': '2023-05-15T11:43:03.561608Z'},
                       {'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                        'available': '0.0',
                        'change': '1.16178',
                        'currency_code': 'LTC',
                        'locked': '0.0',
                        'sequence': 9688664235,
                        'time': '2022-10-12T00:56:19.697459Z'}],
          'channel_name': 'ACCOUNT_HISTORY',
          'time': '2023-05-16T09:45:22.571643Z',
          'type': 'BALANCES_SNAPSHOT'},
 'total': {'LTC': 0.0, 'USDT': 34.144},
 'used': {'LTC': 0.0, 'USDT': 10.090472}}
```

```
p bitpanda watchOrders  
Python v3.10.9
CCXT v3.0.104
bitpanda.watchOrders()
[{'amount': 0.0004,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': '2023-05-16T09:47:16.848Z',
  'fee': None,
  'fees': [],
  'filled': 0.0,
  'id': '19436f1b-42f8-495f-a4f8-34ea8faffba4',
  'info': {'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
           'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
           'activity': 'TRADING',
           'id': '27730db7-9510-4f48-8d6c-765b37b70734',
           'locked': {'amount': '10.090472',
                      'currency_code': 'USDT',
                      'new_available': '24.053528',
                      'new_locked': '10.090472'},
           'order': {'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                     'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
                     'amount': '0.0004',
                     'filled_amount': '0.0',
                     'instrument_code': 'BTC_USDT',
                     'is_post_only': False,
                     'order_id': '19436f1b-42f8-495f-a4f8-34ea8faffba4',
                     'price': '25226.18',
                     'side': 'BUY',
                     'time': '2023-05-16T09:47:16.848499Z',
                     'time_in_force': 'GOOD_TILL_CANCELLED',
                     'type': 'LIMIT'},
           'order_id': '19436f1b-42f8-495f-a4f8-34ea8faffba4',
           'sequence': 11806474984,
           'time': '2023-05-16T09:47:16.853653Z',
           'type': 'ORDER_CREATED'},
  'lastTradeTimestamp': None,
  'postOnly': False,
  'price': 25226.18,
  'reduceOnly': None,
  'remaining': 0.0004,
  'side': 'buy',
  'status': None,
  'stopPrice': None,
  'symbol': 'BTC/USDT',
  'timeInForce': 'GTC',
  'timestamp': 1684230436848,
  'trades': [],
  'triggerPrice': None,
  'type': 'limit'}]
```

```
p bitpanda watchMyTrades      
Python v3.10.9
CCXT v3.0.104
bitpanda.watchMyTrades()
[{'amount': 0.00044,
  'cost': 11.927718,
  'datetime': '2023-05-16T09:56:11.628Z',
  'fee': {'cost': 1.1e-06, 'currency': 'BTC', 'rate': 0.25},
  'fees': [{'cost': 1.1e-06, 'currency': 'BTC', 'rate': 0.25}],
  'id': '2b0438d2-ff78-4541-9efd-45b7401a5ade',
  'info': {'account_holder': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
           'account_id': 'b7f4e27e-b34a-493a-b0d4-4bd341a3f2e0',
           'amount': '0.00044',
           'instrument_code': 'BTC_USDT',
           'order_id': '6dc3dcd7-62e6-4279-a5bc-6d8298e453d3',
           'price': '27108.45',
           'price_tick_sequence': 0,
           'side': 'BUY',
           'time': '2023-05-16T09:56:11.628620Z',
           'trade_id': '2b0438d2-ff78-4541-9efd-45b7401a5ade'},
  'order': '6dc3dcd7-62e6-4279-a5bc-6d8298e453d3',
  'price': 27108.45,
  'side': 'buy',
  'symbol': 'BTC/USDT',
  'takerOrMaker': 'taker',
  'timestamp': 1684230971628,
  'type': None}]

```

```
 p bitpanda watchOHLCV "BTC/USDT"
Python v3.10.9
CCXT v3.0.104
bitpanda.watchOHLCV(BTC/USDT)
[[1684231259999, 27096.39, 27096.39, 27096.26, 27096.26, 32.515603]]
```
